### PR TITLE
Defer distributing index over generic object types

### PR DESCRIPTION
**Pull request by Andarist**
Fixes https://github.com/microsoft/TypeScript/issues/50539

cc @ahejlsberg , the `shouldDeferIndexType` check has been introduced based on the changes in this PR: https://github.com/microsoft/TypeScript/pull/49696/files

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
@typescript-bot test top100

**Comment by ahejlsberg** [comment]
@DanielRosenwasser @sandersn Looks like the top100 test is bombing out in one of the processes. Same issue over in #50535.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15903,6 +15903,9 @@ namespace ts {
         }
 
         function distributeIndexOverObjectType(objectType: Type, indexType: Type, writing: boolean) {
+            if (shouldDeferIndexType(objectType)) {
+                return;
+            }
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:15908]: Remove this.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -15903,6 +15903,9 @@ namespace ts {
         }
 
         function distributeIndexOverObjectType(objectType: Type, indexType: Type, writing: boolean) {
+            if (shouldDeferIndexType(objectType)) {
+                return;
+            }
             // (T | U)[K] -> T[K] | U[K] (reading)
             // (T | U)[K] -> T[K] & U[K] (writing)
             // (T & U)[K] -> T[K] & U[K]
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:15911]: Change the following line to:
```ts
if (objectType.flags & TypeFlags.Union || objectType.Flags & TypeFlags.Intersection && !shouldDeferIndexType(objectType)) ...
```

Comment (ahejlsberg) [reply src/compiler/checker.ts:15911]: I'm concerned that `shouldDeferIndexType` will also do (possibly expensive) work for union types, so I'd prefer only deferring for intersection types.

**Review by ahejlsberg — COMMENTED** [review]
This change makes sense, but I'd prefer reducing the scope a bit as suggested here.

**Comment by Andarist** [comment]
@ahejlsberg I've applied the suggested change and rerun the test suite locally - everything looks OK on this front, somebody from the TS team has to start new perf tests though.

**Comment by jakebailey** [comment]
@typescript-bot perf test this my liege

**Comment by ahejlsberg** [comment]
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot test top100

**Review by ahejlsberg — APPROVED** [review]

**Comment by Andarist** [comment]
@ahejlsberg should this perhaps be cherry-picked into the 4.8 release?

**Comment by ahejlsberg** [comment]
@Andarist Yup, it's already on the list.

**Comment by andrewbranch** [comment]
@typescript-bot cherry-pick this to release-4.8

**Pull request merged**